File: Java/1136. Parallel Courses.java
class Solution {
    public int minimumSemesters(int n, int[][] relations) {
        int[] indegrees = new int[n];
        Map<Integer, List<Integer>> adj = new HashMap<>();
        
        /* record the adjacency info */
        for (int[] edge: relations) {
            if (!adj.containsKey(edge[0])) {
                adj.put(edge[0], new ArrayList<Integer>());
            }
            adj.get(edge[0]).add(edge[1]);
            indegrees[edge[1]-1]++;
        }
        
        Queue<Integer> q = new LinkedList<>();
        /* add all starting courses to the queue */
        for (int i=0; i<n; i++) {
            if (indegrees[i] == 0) {
                q.offer(i+1);
            }
        }
        
        int ans = 0, studied = 0, size;
        while (!q.isEmpty()) {
            ans++;
            Queue<Integer> temp = new LinkedList<>();
            for (int node: q) {
                studied++;
                List<Integer> neis = adj.getOrDefault(node, new ArrayList<Integer>());
                for (int nei: neis) {
                    indegrees[nei-1]--;
                    if (indegrees[nei-1] == 0) {
                        temp.offer(nei);
                    }
                }
            }
            q = temp;
        }
        
        return studied == n ? ans : -1;
    }
}

File: Python3/1136. Parallel Courses.py
class Solution:
    def minimumSemesters(self, n: int, relations: List[List[int]]) -> int:
        indegrees = [0] * n
        adj = collections.defaultdict(list)
        
        # record the adjacency info
        for pre, cur in relations:
            indegrees[cur-1] += 1
            adj[pre].append(cur)
            
        q = deque()
        visited = set()
        
        # add the start nodes
        for node in range(n):
            if indegrees[node] == 0:
                q.append(node+1)
        
        ans = studied = 0
        while q:
            size = len(q)
            ans += 1
            
            for _ in range(size):
                studied += 1
                for nei in adj[q.popleft()]:
                    indegrees[nei-1] -= 1
                    
                    # all pres are learnt
                    if indegrees[nei-1] == 0:
                        q.append(nei)
                
        return ans if studied == n else -1
        